File: tests/fixtures/typescript/sample-repo/src/state.ts
// Mutation-effect patterns.

let counter = 0; // module-scope mutable

export function incrementGlobal(): void {
  // expect: Mut=definite
  counter += 1;
}

export function readGlobal(): number {
  // expect: (none — Pure)
  return counter;
}

export function pushItem<T>(items: T[], item: T): void {
  // expect: Mut=definite
  items.push(item);
}

export function clearList<T>(items: T[]): void {
  // expect: Mut=definite
  items.length = 0;
}

export function patchUser(user: Record<string, unknown>, changes: Record<string, unknown>): void {
  // expect: Mut=definite
  Object.assign(user, changes);
}

export class Counter {
  value = 0;

  increment(): void {
    // expect: Mut=definite
    this.value += 1;
  }
}

export function addOne(x: number): number {
  // expect: Mut=possible
  // Pure body; name starts with "add".
  return x + 1;
}

export function withItem<T>(items: ReadonlyArray<T>, item: T): T[] {
  // expect: (none — Pure)
  // Returns a new array; no mutation, no Mut tag.
  return [...items, item];
}

File: tests/fixtures/rust/sample-repo/src/state.rs
// Mutation-effect patterns.

use std::sync::Mutex;

pub static ERRORS: Mutex<Vec<String>> = Mutex::new(Vec::new());

pub fn record_error(msg: &str) {
    // expect: Mut=definite
    let mut guard = ERRORS.lock().expect("poisoned");
    guard.push(msg.to_string());
}

pub fn append_in_place(items: &mut Vec<i32>, item: i32) {
    // expect: Mut=definite
    items.push(item);
}

pub fn clear_in_place<T>(items: &mut Vec<T>) {
    // expect: Mut=definite
    items.clear();
}

pub struct Counter {
    pub value: i64,
}

impl Counter {
    pub fn new() -> Self {
        // expect: (none — Pure)
        Self { value: 0 }
    }

    pub fn increment(&mut self) {
        // expect: Mut=definite
        self.value += 1;
    }
}

pub fn add_one(x: i64) -> i64 {
    // expect: Mut=possible
    // Pure body; name starts with "add".
    x + 1
}

pub fn with_item<T: Clone>(items: &[T], item: T) -> Vec<T> {
    // expect: (none — Pure)
    let mut out = items.to_vec();
    out.push(item);
    out
}
